### PR TITLE
Add table aws_codebuild_project closes #297

### LINKED ISSUE #297: Add table aws_codebuild_project
**References**
To accomplish the below 2 requirements from AWS Foundational Security Best Practices we need this table.
- CodeBuild GitHub or Bitbucket source repository URLs should use OAuth
- CodeBuild project environment variables should not contain clear text credentials

### PR DESCRIPTION
**Pull request by Paulami30**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_codebuild_project []

PRETEST: tests/aws_codebuild_project

TEST: tests/aws_codebuild_project
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_partition.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_vpc.my_vpc: Creating...
aws_iam_role.test_role: Creating...
aws_s3_bucket.test_bucket: Creating...
aws_s3_bucket.test_bucket: Still creating... [10s elapsed]
aws_iam_role.test_role: Still creating... [10s elapsed]
aws_vpc.my_vpc: Still creating... [10s elapsed]
aws_vpc.my_vpc: Still creating... [20s elapsed]
aws_s3_bucket.test_bucket: Still creating... [20s elapsed]
aws_iam_role.test_role: Still creating... [20s elapsed]
aws_iam_role.test_role: Creation complete after 26s [id=turbottest48279]
aws_s3_bucket.test_bucket: Still creating... [30s elapsed]
aws_vpc.my_vpc: Still creating... [30s elapsed]
aws_s3_bucket.test_bucket: Still creating... [40s elapsed]
aws_vpc.my_vpc: Still creating... [40s elapsed]
aws_vpc.my_vpc: Still creating... [50s elapsed]
aws_s3_bucket.test_bucket: Still creating... [50s elapsed]
aws_s3_bucket.test_bucket: Still creating... [1m0s elapsed]
aws_vpc.my_vpc: Still creating... [1m0s elapsed]
aws_vpc.my_vpc: Still creating... [1m10s elapsed]
aws_s3_bucket.test_bucket: Still creating... [1m10s elapsed]
aws_vpc.my_vpc: Creation complete after 1m11s [id=vpc-007a66fa8dfd15e49]
aws_subnet.my_subnet2: Creating...
aws_subnet.my_subnet1: Creating...
aws_s3_bucket.test_bucket: Still creating... [1m20s elapsed]
aws_subnet.my_subnet1: Still creating... [10s elapsed]
aws_subnet.my_subnet2: Still creating... [10s elapsed]
aws_s3_bucket.test_bucket: Still creating... [1m30s elapsed]
aws_subnet.my_subnet1: Still creating... [20s elapsed]
aws_subnet.my_subnet2: Still creating... [20s elapsed]
aws_subnet.my_subnet1: Creation complete after 20s [id=subnet-02f2fdf011ab7815e]
aws_subnet.my_subnet2: Creation complete after 20s [id=subnet-0f05f70f5231dce27]
aws_s3_bucket.test_bucket: Creation complete after 1m38s [id=turbottest48279]
aws_iam_role_policy.test_role_policy: Creating...
aws_codebuild_project.named_test_resource: Creating...
aws_iam_role_policy.test_role_policy: Still creating... [10s elapsed]
aws_codebuild_project.named_test_resource: Still creating... [10s elapsed]
aws_iam_role_policy.test_role_policy: Creation complete after 13s [id=turbottest48279:terraform-20210415110038971200000001]
aws_codebuild_project.named_test_resource: Creation complete after 14s [id=arn:aws:codebuild:us-east-1:123456789012:project/turbottest48279]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Warning: Interpolation-only expressions are deprecated

  on variables.tf line 155, in resource "aws_codebuild_project" "named_test_resource":
 155:     location = "${aws_s3_bucket.test_bucket.bucket}"

Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.

Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.


Apply complete! Resources: 7 added, 0 changed, 0 destroyed.

Outputs:

resource_aka = arn:aws:codebuild:us-east-1:123456789012:project/turbottest48279
resource_id = arn:aws:codebuild:us-east-1:123456789012:project/turbottest48279
resource_name = turbottest48279

Running SQL query: test-get-query.sql
[
  {
    "artifacts": {
      "ArtifactIdentifier": null,
      "EncryptionDisabled": null,
      "Location": null,
      "Name": null,
      "NamespaceType": null,
      "OverrideArtifactName": false,
      "Packaging": null,
      "Path": null,
      "Type": "NO_ARTIFACTS"
    },
    "cache": {
      "Location": "turbottest48279",
      "Modes": null,
      "Type": "S3"
    },
    "description": "Resource for testing",
    "name": "turbottest48279",
    "source": {
      "Auth": null,
      "BuildStatusConfig": null,
      "Buildspec": "",
      "GitCloneDepth": 0,
      "GitSubmodulesConfig": null,
      "InsecureSsl": false,
      "Location": "https://github.com/mitchellh/packer.git",
      "ReportBuildStatus": false,
      "SourceIdentifier": null,
      "Type": "GITHUB"
    },
    "title": "turbottest48279"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:codebuild:us-east-1:123456789012:project/turbottest48279"
    ],
    "description": "Resource for testing",
    "name": "turbottest48279",
    "tags": {
      "foo": "bar"
    },
    "title": "turbottest48279"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "artifacts": {
      "ArtifactIdentifier": null,
      "EncryptionDisabled": null,
      "Location": null,
      "Name": null,
      "NamespaceType": null,
      "OverrideArtifactName": false,
      "Packaging": null,
      "Path": null,
      "Type": "NO_ARTIFACTS"
    },
    "description": "Resource for testing",
    "name": "turbottest48279",
    "tags": {
      "foo": "bar"
    },
    "title": "turbottest48279"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

POSTTEST: tests/aws_codebuild_project

TEARDOWN: tests/aws_codebuild_project

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

> select
  name,
  description,
  encryption_key,
  concurrent_build_limit,
  source_version,
  service_role,
  created,
  last_modified,
  region
from
  aws_codebuild_project;
+----------+-------------+-------------------------------------------------+------------------------+----------------+----------------------------------------
| name     | description | encryption_key                                  | concurrent_build_limit | source_version | service_role                           
+----------+-------------+-------------------------------------------------+------------------------+----------------+----------------------------------------
| test-001 | testing     | arn:aws:kms:us-east-1:123456789012:alias/aws/s3 | 2                      | 2              | arn:aws:iam::123456789012:role/service-
+----------+-------------+-------------------------------------------------+------------------------+----------------+----------------------------------------

### Get the build input details of the projects

> select
  name,
  source_version,
  source ->> 'Auth' as auth,
  source ->> 'BuildStatusConfig' as build_status_config,
  source ->> 'Buildspec' as build_spec,
  source ->> 'GitCloneDepth' as git_clone_depth,
  source ->> 'GitSubmodulesConfig' as git_submodules_config,
  source ->> 'InsecureSsl' as insecure_ssl,
  source ->> 'Location' as location,
  source ->> 'ReportBuildStatus' as report_build_status,
  source ->> 'SourceIdentifier' as source_identifier,
  source ->> 'Type' as type
from
  aws_codebuild_project;
+----------+----------------+--------+---------------------+------------+-----------------+-----------------------+--------------+--------------------+-------
| name     | source_version | auth   | build_status_config | build_spec | git_clone_depth | git_submodules_config | insecure_ssl | location           | report
+----------+----------------+--------+---------------------+------------+-----------------+-----------------------+--------------+--------------------+-------
| test-001 | 2              | <null> | <null>              | <null>     | <null>          | <null>                | false        | nlbbucpaul/test123 | <null>
+----------+----------------+--------+---------------------+------------+-----------------+-----------------------+--------------+--------------------+-------

### List of projects which are not created within a VPC

> select
  name,
  description,
  vpc_config
from
  aws_codebuild_project
where
  vpc_config is null;
+----------+-------------+------------+
| name     | description | vpc_config |
+----------+-------------+------------+
| test-001 | testing     | <null>     |
+----------+-------------+------------+

### List of projects in which logging is not enabled

> select
  name,
  description,
  logs_config -> 'CloudWatchLogs' ->> 'Status' as cloud_watch_logs_status,
  logs_config -> 'S3Logs' ->> 'Status' as s3_logs_status
from
  aws_codebuild_project
where
  logs_config -> 'CloudWatchLogs' ->> 'Status' = 'DISABLED'
  and logs_config -> 'S3Logs' ->> 'Status' = 'DISABLED';
+----------+-------------+-------------------------+----------------+
| name     | description | cloud_watch_logs_status | s3_logs_status |
+----------+-------------+-------------------------+----------------+
| test-001 | testing     | DISABLED                | DISABLED       |
+----------+-------------+-------------------------+----------------+
```
</details>

### CODE REVIEW

File: aws/service.go [hunk 1]
```diff
@@ -159,6 +160,26 @@ func BackupService(ctx context.Context, d *plugin.QueryData, region string) (*ba
 	return svc, nil
 }
 
+// CodeBuildService returns the service connection for AWS CodeBuild service
+func CodeBuildService(ctx context.Context, d *plugin.QueryData, region string) (*codebuild.CodeBuild, error) {
+	if region == "" {
+		return nil, fmt.Errorf("region must be passed BackupService")
```
Comment (Subhajit97) [inline aws/service.go:166]:
```suggestion
		return nil, fmt.Errorf("region must be passed CodeBuildService")
```

File: aws/service.go [hunk 2]
```diff
@@ -159,6 +160,26 @@ func BackupService(ctx context.Context, d *plugin.QueryData, region string) (*ba
...
+	}
+	// have we already created and cached the service?
+	serviceCacheKey := fmt.Sprintf("backup-%s", region)
```
Comment (Subhajit97) [inline aws/service.go:169]:
```suggestion
	serviceCacheKey := fmt.Sprintf("codebuild-%s", region)
```

File: aws/table_aws_codebuild_project.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+	if op.Projects != nil && len(op.Projects) > 0 {
+		return op.Projects[0], nil
+	}
+	return nil, nil
+}
+
+//// TRANSFORM FUNCTIONS
+
+func getCodeBuildProjectTurbotTags(ctx context.Context, d *transform.TransformData) (interface{},
```
Comment (Subhajit97) [inline aws/table_aws_codebuild_project.go:283]:
```suggestion
func codeBuildProjectTurbotTags(ctx context.Context, d *transform.TransformData) (interface{},
```

File: docs/tables/aws_codebuild_project.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  source ->> 'Location' as location,
+  source ->> 'ReportBuildStatus' as report_build_status,
+  source ->> 'SourceIdentifier' as source_identifier,
+  source ->> 'Type' as type
+from
+  aws_codebuild_project;
+```
+
+
+### List of projects which are not created within a VPC
```
Comment (Subhajit97) [inline docs/tables/aws_codebuild_project.md:46]:
```suggestion
### List projects which are not created within a VPC
```

File: docs/tables/aws_codebuild_project.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  description,
+  vpc_config
+from
+  aws_codebuild_project
+where
+  vpc_config is null;
+```
+
+
+### List of projects in which logging is not enabled
```
Comment (Subhajit97) [inline docs/tables/aws_codebuild_project.md:60]:
```suggestion
### List projects with logging disabled
```

File: aws-test/tests/aws_codebuild_project/variables.tf [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+resource "aws_codebuild_project" "named_test_resource" {
+  name         = var.resource_name
+  description  = "Resource for testing"
+  service_role = aws_iam_role.test_role.arn
+  artifacts {
+    type = "NO_ARTIFACTS"
+  }
+  cache {
+    type     = "S3"
+    location = "${aws_s3_bucket.test_bucket.bucket}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_codebuild_project/variables.tf:155]:
```suggestion
    location = aws_s3_bucket.test_bucket.bucket
```

File: aws-test/tests/aws_codebuild_project/variables.tf [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+data "null_data_source" "resource" {
+  inputs = {
+    scope = "arn:${data.aws_partition.current.partition}:::${data.aws_caller_identity.current.account_id}"
+  }
+}
+
+resource "aws_vpc" "my_vpc" {
+  cidr_block = "10.1.0.0/16"
+}
+resource "aws_subnet" "my_subnet1" {
```
Comment (Subhajit97) [inline aws-test/tests/aws_codebuild_project/variables.tf:53]: Add an extra line between plans

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
@Paulami30 There are few changes required. Please see comments, thanks!

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**